### PR TITLE
Removal of additionalAccountFields & additionalInfo from cache items

### PR DESCRIPTION
**Pull request by d-col**

### CODE REVIEW

File: IdentityCore/src/cache/token/MSIDLegacyTokenCacheItem.m [hunk 1]
```diff
@@ -123,12 +126,19 @@ - (instancetype)initWithCoder:(NSCoder *)coder
...
     {
-        self.extendedExpiresOn = self.additionalInfo[MSID_EXTENDED_EXPIRES_ON_CACHE_KEY];
-        NSMutableDictionary *additionalServer = [[NSMutableDictionary alloc] initWithDictionary:self.additionalInfo];
+        self.extendedExpiresOn = additionalServer[MSID_EXTENDED_EXPIRES_ON_CACHE_KEY];
         [additionalServer removeObjectForKey:MSID_EXTENDED_EXPIRES_ON_CACHE_KEY];
+    }
+    if(additionalServer[MSID_SPE_INFO_CACHE_KEY])
+    {
+        self.speInfo = additionalServer[MSID_SPE_INFO_CACHE_KEY];
+        [additionalServer removeObjectForKey:MSID_SPE_INFO_CACHE_KEY];
```
Comment (oldalton) [inline IdentityCore/src/cache/token/MSIDLegacyTokenCacheItem.m]: I think ADAL is just looking into additional info. Setting spe info separately would require updates in ADAL.

File: IdentityCore/src/cache/token/MSIDLegacyTokenCacheItem.m [hunk 2]
```diff
@@ -173,16 +183,16 @@ - (void)encodeWithCoder:(NSCoder *)coder
 
     [coder encodeObject:[NSMutableDictionary dictionary] forKey:@"additionalClient"];
 
+    NSMutableDictionary* additionalServer = [[NSMutableDictionary alloc] initWithDictionary:self.additionalInfo];
```
Comment (oldalton) [inline IdentityCore/src/cache/token/MSIDLegacyTokenCacheItem.m]: nit: code style, should be NSMutableDictionary *additionalServer

File: IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m [hunk 3]
```diff
@@ -263,7 +263,9 @@ - (BOOL)saveAccount:(nonnull MSIDAccountCacheItem *)account
     if (previousAccount)
     {
         // Make sure we copy over all the additional fields
-        [account updateFieldsFromAccount:previousAccount];
+        NSMutableDictionary *mergedDictionary = [previousAccount.jsonDictionary mutableCopy];
+        [mergedDictionary addEntriesFromDictionary:account.jsonDictionary];
+        account = [[MSIDAccountCacheItem alloc] initWithJSONDictionary:mergedDictionary error:error];
```
Comment (oldalton) [inline IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m]: nit: check that account creation succeeded before proceeding with saving

File: IdentityCore/src/oauth2/token/MSIDAccessToken.m [hunk 4]
```diff
@@ -109,6 +109,12 @@ - (instancetype)initWithTokenCacheItem:(MSIDCredentialCacheItem *)tokenCacheItem
     {
         _expiresOn = tokenCacheItem.expiresOn;
         _extendedExpiresOn = tokenCacheItem.extendedExpiresOn;
+        NSMutableDictionary *serverInfo = [NSMutableDictionary dictionaryWithDictionary:_additionalServerInfo];
+        [serverInfo removeObjectForKey:MSID_EXTENDED_EXPIRES_ON_CACHE_KEY];
```
Comment (oldalton) [inline IdentityCore/src/oauth2/token/MSIDAccessToken.m]: why is this necessary?

File: IdentityCore/src/oauth2/token/MSIDBaseToken.m [hunk 5]
```diff
@@ -139,9 +139,21 @@ - (instancetype)initWithTokenCacheItem:(MSIDCredentialCacheItem *)tokenCacheItem
             MSID_LOG_WARN(nil, @"Trying to initialize token when missing clientId field");
             return nil;
         }
-        
-        _additionalServerInfo = tokenCacheItem.additionalInfo;
 
+        NSMutableDictionary* serverInfo = [NSMutableDictionary new];
```
Comment (oldalton) [inline IdentityCore/src/oauth2/token/MSIDBaseToken.m]: nit: would it make it easier if we handled those fields similarly between cache items and tokens? I'm fine getting rid of additionalInfo outside of legacy cache item serialization and additional fields coming from server that we don't understand.

File: IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m (hunk 3 above)
Comment (d-col) [reply IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m]: updated

File: IdentityCore/src/oauth2/token/MSIDAccessToken.m (hunk 4 above)
Comment (d-col) [reply IdentityCore/src/oauth2/token/MSIDAccessToken.m]: The base class sets extendedExpiresOn in the dictionary, while MSIDAccessToken has a specific property for this. Having the value in both places is ambiguous.

File: IdentityCore/src/oauth2/token/MSIDBaseToken.m (hunk 5 above)
Comment (d-col) [reply IdentityCore/src/oauth2/token/MSIDBaseToken.m]: Seems like a good idea. Updated.

File: IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m [hunk 6]
```diff
@@ -259,7 +259,18 @@ - (BOOL)saveAccount:(nonnull MSIDAccountCacheItem *)account
     if (previousAccount)
     {
         // Make sure we copy over all the additional fields
-        [account updateFieldsFromAccount:previousAccount];
+        NSMutableDictionary *mergedDictionary = [previousAccount.jsonDictionary mutableCopy];
+        [mergedDictionary addEntriesFromDictionary:account.jsonDictionary];
+        NSError *accountError;
+        account = [[MSIDAccountCacheItem alloc] initWithJSONDictionary:mergedDictionary error:&accountError];
+        if (accountError != nil || account == nil)
```
Comment (oldalton) [inline IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m]: nit: you can just check for !account here

File: IdentityCore/src/oauth2/token/MSIDBaseToken.h [hunk 7]
```diff
@@ -56,6 +56,8 @@
 @property (readwrite) NSString *clientId;
 @property (readwrite) NSDictionary *additionalServerInfo;
 @property (readwrite) MSIDAccountIdentifier *accountIdentifier;
+@property (readwrite) NSDate *extendedExpiresOn;
```
Comment (oldalton) [inline IdentityCore/src/oauth2/token/MSIDBaseToken.h]: I don't think this one is applicable to base token. it should be in access token only similarly to expiresOn.

File: IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m (hunk 6 above)
Comment (d-col) [reply IdentityCore/src/cache/accessor/MSIDAccountCredentialCache.m]: updated

File: IdentityCore/src/oauth2/token/MSIDBaseToken.h (hunk 7 above)
Comment (d-col) [reply IdentityCore/src/oauth2/token/MSIDBaseToken.h]: updated
